### DISCUSSION TITLE
Add optional feature to remove unused imports

### DISCUSSION THREAD

**Discussion by sweepline** [eslint/eslint#18494] [Ideas]
Hi, im the maintainer of [eslint-plugin-unused-imports](https://github.com/sweepline/eslint-plugin-unused-imports). I have started switching projects over to [biome](https://biomejs.dev/) because they have a nice optional fixer for removing unused imports. They dont guarantee that fixing these wont break side-effects, and neither does my package.

My package is quite heavily used, so I think that shows its something people want even if it might break side-effect imports. 
I was wondering if the functionality could be upstreamed here in some way, as my plugin does monkey-patching to make it work and its quite fragile with regards to updates. It could be an opt-in feature for the people that want it?

Regards, Mikkel.

**Comment by fasttime** [comment]
Hi @sweepline, thanks for the suggestion. Autofixes in ESLint are always provided by rules, and built-in rules are not allowed to apply autofixes if there is a chance that doing so will change the semantics of the code because of side effects. This is to prevent introducing accidental bugs that could be hard to track. Instead, a rule can [provide suggestions](https://eslint.org/docs/latest/extend/custom-rules#providing-suggestions) to offer this kind of changes to a user.

On the other hand it's perfectly fine that third-party plugins like yours provide this kind of functionality, and I'd encourage you to keep maintaining that plugin since it's become so popular. If you can explain why you need monkey-patching to make your plugin work, maybe we can try to find a more robust solution that works for your use case.

**Reply by sweepline** [reply]
Hi, and thanks for your reply. I understand the sentiment, I just thought that if its opt-in then maybe it would be alright to have upstream.

The reason i have ended up monkey patching is because I actually use the built-in no-unused-vars rule and then split it up into two rules. One rule that has an autofixer for removing said variable and one that is just the built-in rule. This is split up based on the AST type being "ImportSpecifier", "ImportDefaultSpecifier" or "ImportNamespaceSpecifier".
It would be awesome if there was a way to augment the built-in rule without resolving to monkey-patching the rules, or some other way of achieving this without copying the entire source for the no-unused-vars rule.

**Reply by fasttime** [reply]
I see. The general contract of ESLint rules is quite stable with only sporadic changes delivered during major releases, but it's true that the specifics of each built-in rule (like the listeners on AST nodes, the problems reported, the autofix behavior, etc.) can change anytime. That said, it's not uncommon for rules to encapsulate other rules, there are even libraries to do that, and if the underlying rule is changed, things can break unexpectedly.

>  It would be awesome if there was a way to augment the built-in rule without resolving to monkey-patching the rules, or some other way of achieving this without copying the entire source for the no-unused-vars rule.

There's nothing like that at the moment, but if you have an idea on how the augmentation would work, feel free to [submit a change request](https://eslint.org/docs/latest/contribute/request-change).
